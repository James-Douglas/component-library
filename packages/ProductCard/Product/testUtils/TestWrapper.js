import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
