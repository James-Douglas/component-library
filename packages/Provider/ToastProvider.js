import React, { useState } from 'react';
import PropTypes from 'prop-types';
import ToastContext from './contexts/Toast/toast.context';

const uuidv4 = require('uuid/v4');

const ToastProvider = ({ children }) => {
  const [toasts, setToasts] = useState([]);

  const addToast = (toast) => {
    const id = uuidv4();
    const clonedToast = React.cloneElement(toast, { id });
    setToasts([...toasts, clonedToast]);
    return id;
  };

  const removeToast = (id) => {
    setToasts((prevState) => prevState.filter((n) => n.props.id !== id));
  };

  const toastInfo = { addToast, removeToast, toasts };

  return (
    <ToastContext.Provider value={toastInfo}>
      {children}
    </ToastContext.Provider>
  );
};

ToastProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]).isRequired,
};

export default ToastProvider;
