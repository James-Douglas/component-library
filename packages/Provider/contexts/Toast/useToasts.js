import { useContext } from 'react';
import ToastContext from './toast.context';

const useToasts = () => useContext(ToastContext);

export default useToasts;
