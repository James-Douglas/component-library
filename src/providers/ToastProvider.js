import React, { useState } from 'react';
import PropTypes from 'prop-types';

import ToastContext from '../contexts/Toast/toast.context';

const uuidv4 = require('uuid/v4');

const ToastProvider = ({ children }) => {
  const [toasts, setToasts] = useState([]);

  const addToast = (props) => {
    const id = uuidv4();
    const toast = { props, id };
    setToasts([...toasts, toast]);
    return id;
  };

  const removeToast = (id) => {
    setToasts((prevState) => prevState.filter((n) => n.id !== id));
  };

  const toastInfo = { addToast, removeToast, toasts };

  return (
    <ToastContext.Provider value={toastInfo}>
      {children}
    </ToastContext.Provider>
  );
};

ToastProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]).isRequired,
};

export default ToastProvider;
