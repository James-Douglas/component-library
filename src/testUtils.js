// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
