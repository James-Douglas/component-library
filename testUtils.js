import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
