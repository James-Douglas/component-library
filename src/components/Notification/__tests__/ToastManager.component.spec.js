import React from 'react';
import ReactDOM from 'react-dom';
import { ThemeProvider } from 'styled-components';
import { renderHook, act } from '@testing-library/react-hooks';
import theme from '../../../themes/ctm.theme';
import { render } from '../../../testUtils';
import ToastManager from '../ToastManager.component';
import ToastProvider from '../../../providers/ToastProvider';
import useToasts from '../../../contexts/Toast/useToasts';
import 'jest-styled-components';

describe('ToastManager()', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element, node) => element);
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  it('renders a container ', () => {
    const { container } = render(
      <ToastManager />,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('adds a toast to the toasts state', () => {
    // eslint-disable-next-line react/prop-types
    const wrapper = ({ children }) => (
      <ThemeProvider theme={theme}>
        <ToastProvider>
          <ToastManager />
          {children}
        </ToastProvider>
      </ThemeProvider>
    );
    const { result } = renderHook(() => useToasts(), { wrapper });

    act(() => {
      result.current.addToast(
        {
          type: 'toast',
          variant: 'general',
          title: 'General Toast Message',
          content: 'This will auto dissmiss in 5 seconds',
          primaryAction: {
            content: 'Action goes here',
            link: '#',
          },
          icon: true,
          closeButton: true,
          autoClose: 5000,
        },
      );
    });
    const toastArray = result.current.toasts;

    expect(toastArray).toHaveLength(1);
    expect(result.current.toasts[0].props.type).toBe('toast');
    expect(result.current.toasts[0].props.variant).toBe('general');
    expect(result.current.toasts[0].props.title).toBe('General Toast Message');
  });

  it('removes a toast from the toasts state', () => {
    // eslint-disable-next-line react/prop-types
    const wrapper = ({ children }) => (
      <ThemeProvider theme={theme}>
        <ToastProvider>
          <ToastManager />
          {children}
        </ToastProvider>
      </ThemeProvider>
    );
    const { result } = renderHook(() => useToasts(), { wrapper });

    let toastID = '';
    act(() => {
      toastID = result.current.addToast(
        {
          type: 'toast',
          variant: 'general',
          title: 'General Toast Message',
          content: 'This will auto dissmiss in 5 seconds',
          primaryAction: {
            content: 'Action goes here',
            link: '#',
          },
          icon: true,
          closeButton: true,
          autoClose: 5000,
        },
      );
    });

    act(() => {
      result.current.removeToast(toastID);
    });
    const toastArray = result.current.toasts;
    expect(toastArray).toHaveLength(0);
  });
});
