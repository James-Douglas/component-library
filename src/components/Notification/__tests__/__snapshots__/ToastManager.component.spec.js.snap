// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastManager() renders a container  1`] = `"<div class=\\"sc-fzozJi Wfmzj\\"><div class=\\"sc-fzoLsD dOyZm\\"></div></div>"`;
